### PR TITLE
Remove type subclassing

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #24284 Merge ProfiledTensorType and TensorType
* **#24257 Remove type subclassing**
* #24169 remove CompleteTensorType

Type subclassing was used to support our old hierarchy of
Tensor types. Now that we have one tensor type it is not needed.
This removes:

* isSubclass, since it is now always false.
* type slicing, which was only needed for subclasses.
* AutogradZeroTensor, which is folded into ProfiledTensorType

Differential Revision: [D16794035](https://our.internmc.facebook.com/intern/diff/D16794035)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -8208,6 +8208,7 @@ def tensor_unifying(x, y, z):
         c = torch.zeros(2, 4, dtype=torch.float)
 
         graph = _propagate_shapes(tensor_unifying.graph, (a, b, c), False)
+        print(graph)
```
Comment (Krovatkin) [inline test/test_jit.py]: a stray print?

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -614,19 +542,32 @@ struct CAFFE2_API ProfiledTensorType : public TensorType {
...
 
-  ProfiledTensorTypePtr merge(ProfiledTensorTypePtr other) {
+  ProfiledTensorTypePtr merge(ProfiledTensorTypePtr other) const {
     auto scalar_type = merge_primitive(scalarType(), other->scalarType());
     auto dev = merge_primitive(device(), other->device());
     auto sz = sizes().merge(other->sizes());
     auto srs = strides().merge(other->strides());
     auto gr = merge_primitive(requiresGrad(), other->requiresGrad());
-    return ProfiledTensorType::create(scalar_type, dev, sz, srs, gr);
+    auto zero = merge_primitive(autogradZero(), other->autogradZero());
```
Comment (Krovatkin) [inline aten/src/ATen/core/jit_type.h]: does switching from an undefined to defined tensor make sense? Maybe, it should be an assert?
`TORCH_INTERNAL_ASSERT(autogradZero() == other->autogradZero())`
I actually went the same path and added the same field in my branch, but I wasn't sure if it makes sense for a tensor to switch.

File: aten/src/ATen/core/jit_type.h [hunk 3]
```diff
@@ -676,6 +619,12 @@ struct CAFFE2_API ProfiledTensorType : public TensorType {
   VaryingShape sizes_;
   VaryingStrides strides_;
   c10::optional<bool> requires_grad_;
+  // we exploit the fact certain tensors must be zero in the autograd to 
+  // optimize gradient computation. If true, this means that this tensor
+  // must only contain zeros. Normally this will be nullopt, meaning
+  // the tensor may or may not contain only zeros. If false, 
+  // this means the tensor must have some non-zero elements.
+  c10::optional<bool> autograd_zero_;
```
Comment (Krovatkin) [inline aten/src/ATen/core/jit_type.h]: I wish we had this comment before, took me a while to figure out how all these pieces fit together 😢

File: torch/csrc/jit/graph_executor_impl.h [hunk 4]
```diff
@@ -87,7 +87,7 @@ struct GraphExecutorImplBase {
     auto local_graph = this->graph->copy();
     for (size_t i = 0; i < input_values.size(); ++i) {
       // propagate tensor types
-      if (input_values.at(i)->type()->cast<TensorType>()) {
+      if (input_values.at(i)->type()->isSubtypeOf(TensorType::get())) {
```
Comment (Krovatkin) [inline torch/csrc/jit/graph_executor_impl.h]: what's wrong with cast to `TensorType`?

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: aten/src/ATen/core/jit_type.h (hunk 2 above)
Comment (zdevito) [reply aten/src/ATen/core/jit_type.h]: In this case it is merging a tensor which is known to be zero with one that that may/may not be zero. So the result is a tensor which may/may not be zero, which is what this implements.

**Pull request closed without merging**